### PR TITLE
PT-161750304 upstream sync cuckoo

### PR DESCRIPTION
**Pull request by hanssv**
Do an upstream sync with tromp/cuckoo.git
    
    Change naming of miners to lean<edge_bits> instead of lean<node_bits>
    to align better with upstream (and cause less confusion).

### CODE REVIEW

**Comment by hanssv** [comment]
https://www.pivotaltracker.com/story/show/161750304

Should be ok now - will take a look at `protocol` and see if we mention `NODE_SIZE`/`EDGE_SIZE` there...

File: system_test/only_local/aest_heavy_sync_SUITE_data/epoch.yaml.mustache [hunk 1]
```diff
@@ -31,7 +31,7 @@ mining:
     autostart: {{mining.autostart}}
     cuckoo:
         miner:
-            executable: mean28s-generic
-            extra_args: ""
-            node_bits: 28
+            executable: mean29-generic
+            extra_args: "-t 5"
+            edge_bits: 29
```
Comment (UlfNorell) [inline system_test/only_local/aest_heavy_sync_SUITE_data/epoch.yaml.mustache]: I see there is no `mean27`. Running five threads means test won't be too much slower?

File: docs/release-notes/next/PT-161750304-upstream_sync_cuckoo.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+* Changes naming scheme for miner executables. Used to be `lean/mean/cuda<node-bits>`, this is changed to
+  `lean/mean/cuds<edge-bits>` to align with upstream cuckoo. An upstream sync is also performed. *NOTE:*
```
Comment (lucafavatella) [inline docs/release-notes/next/PT-161750304-upstream_sync_cuckoo.md]: s/cuds/cuda

File: apps/aeutils/priv/epoch_config_schema.json [hunk 3]
```diff
@@ -391,21 +391,21 @@
                             "required" : [
                                 "executable",
                                 "extra_args",
-                                "node_bits"
+                                "edge_bits"
                             ],
                             "properties" : {
                                 "executable" : {
-                                    "description" : "Executable binary of the miner. Options are: \"mean30s-generic\" (memory-intensive), \"mean30s-avx2\" (memory-intensive, benefits from faster CPU supporting AVX2 instructions), \"lean30\" (CPU-intensive, useful if memory-constrained).",
+                                    "description" : "Executable binary of the miner. Options are: \"mean29-generic\" (memory-intensive), \"mean29-avx2\" (memory-intensive, benefits from faster CPU supporting AVX2 instructions), \"lean29-generic\" (CPU-intensive, useful if memory-constrained), \"lean29-avx2\" (CPU-intensive, useful if memory-constrained, benefits from faster CPU supporting AVX2 instructions).",
```
Comment (lucafavatella) [inline apps/aeutils/priv/epoch_config_schema.json]: Will require update to https://github.com/aeternity/epoch/wiki/Troubleshooting#node-failing-mining-attempts

File: apps/aecuckoo/test/aecuckoo_SUITE.erl [hunk 4]
```diff
@@ -21,59 +21,46 @@
...
+    [{miner, 'mean15-generic'} | Config];
+init_per_group(lean15, Config) ->
+    [{miner, 'lean15-generic'} | Config].
 
 end_per_group(_Group, _Config) ->
     ok.
 
 smoke_test(Config) ->
     Nonce = ?config(nonce, Config),
     ExpectedVerifierCycleHash = ?config(cyclehash, Config),
```
Comment (lucafavatella) [inline apps/aecuckoo/test/aecuckoo_SUITE.erl]: Unused?

File: apps/aecuckoo/test/aecuckoo_SUITE.erl [hunk 5]
```diff
@@ -21,59 +21,46 @@
...
-init_per_group(smoke_tests_16, Config) ->
+init_per_group(smoke_tests_15, Config) ->
     [{nonce, 86},
-     {cyclehash, "c6bf28762af60c78600203ea018de966ebee3f37d84ea21f07b114553fe1692f"},
-     {verifier, verify16} | Config];
-init_per_group(mean16, Config) ->
-    [{miner, 'mean16s-generic'} | Config];
-init_per_group(lean16, Config) ->
-    [{miner, 'lean16'} | Config].
+     {cyclehash, "c6bf28762af60c78600203ea018de966ebee3f37d84ea21f07b114553fe1692f"}
```
Comment (lucafavatella) [inline apps/aecuckoo/test/aecuckoo_SUITE.erl]: Unused?

File: apps/aecuckoo/Makefile [hunk 6]
```diff
@@ -1,101 +1,92 @@
...
+
+PRIV = priv/bin
+CUCKOO = c_src/src/cuckoo
+
+HDRS=$(CUCKOO)/cuckoo.h $(CUCKOO)/../crypto/siphash.h
+
+# Flags from upstream makefile
+OPT ?= -O3
+
+GPP_ARCH_FLAGS ?= -m64 -x c++ -march=native
```
Comment (lucafavatella) [inline apps/aecuckoo/Makefile]: I am not sure about `-march=native` option.

Comment (lucafavatella) [reply apps/aecuckoo/Makefile]: Usage of `-march=native` causes all CPU operations available to the machine that is compiling to be used in the compiled binary hence making the resulting binary potentially not able to be supported on other machines. This seems reverting f49528b0 . Could you please clarify.

----

This PR:
* Builds package passing `-m64 ... -march=native`- see [CI](https://circleci.com/gh/aeternity/epoch/40936):
```
(cd c_src/src/cuckoo; g++ -m64 -x c++ -march=native -std=c++11 -Wall -Wno-format -Wno-deprecated-declarations -D_POSIX_C_SOURCE=200112L -O3 -DPREFETCH -I.  -pthread -o mean29-generic -DSAVEEDGES -DNSIPHASH=1 -DEDGEBITS=29 mean.cpp ../crypto/blake2b-ref.c)
```

Master corresponding to this PR:
* Builds package passing only `-m64` (not `-march`) - see [CI](https://circleci.com/gh/aeternity/epoch/40713):
```
gcc -m64 -std=gnu11 -O -fPIC -shared -o libblake2b.so blake2b-ref.c
...
g++ -m64 -std=c++11 -Wall -Wno-format -Wno-deprecated-declarations -D_POSIX_C_SOURCE=200112L -O3 -DPREFETCH -I.  -pthread -o mean30sx1 -DSAVEEDGES -DNSIPHASH=1 -DEDGEBITS=29 mean_miner.cpp -L. -lblake2b
```

File: apps/aecuckoo/test/aecuckoo_SUITE.erl (hunk 4 above)
Comment (hanssv) [reply apps/aecuckoo/test/aecuckoo_SUITE.erl]: No, I don't think so...

File: apps/aecuckoo/Makefile (hunk 6 above)
Comment (hanssv) [reply apps/aecuckoo/Makefile]: Wasn't careful enough when merging the two Makefiles...

**Comment by hanssv** [comment]
Updated - addressed comments by Luca. And fixed some noisy but harmless error messages in the Makefile.

**Comment by hanssv** [comment]
Let's wait for @dincho to also try the GPU version...

**Comment by hanssv** [comment]
Added cuda targets in "our" Makefile to simplify cuda instructions...

**Pull request merged**